# Problem Set 3

## Question-1

Step 1. Import the necessary libraries

In [24]:
import pandas as pd

Step 2. Import the dataset from this address.


In [25]:
users = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user", sep = "|")
users

user_id  age gender     occupation zip_code
0          1   24      M     technician    85711
1          2   53      F          other    94043
2          3   23      M         writer    32067
3          4   24      M     technician    43537
4          5   33      F          other    15213
..       ...  ...    ...            ...      ...
938      939   26      F        student    33319
939      940   32      M  administrator    02215
940      941   20      M        student    97229
941      942   48      F      librarian    78209
942      943   22      M        student    77841

[943 rows x 5 columns]

Step 4. Discover what is the mean age per occupation

In [4]:
users.groupby("occupation").age.mean()


occupation
administrator    38.746835
artist           31.392857
doctor           43.571429
educator         42.010526
engineer         36.388060
entertainment    29.222222
executive        38.718750
healthcare       41.562500
homemaker        32.571429
lawyer           36.750000
librarian        40.000000
marketing        37.615385
none             26.555556
other            34.523810
programmer       33.121212
retired          63.071429
salesman         35.666667
scientist        35.548387
student          22.081633
technician       33.148148
writer           36.311111
Name: age, dtype: float64

In [5]:
users.rename(columns={"sex":"gender"}, inplace=True)

Step 5. Discover the Male ratio per occupation and sort it from the most to the least

In [7]:
male = users.groupby("occupation")["gender"].value_counts(normalize=True)*100
male.head()

occupation     gender
administrator  M          54.430380
               F          45.569620
artist         M          53.571429
               F          46.428571
doctor         M         100.000000
Name: proportion, dtype: float64

In [8]:
gendersn = pd.DataFrame(male)
gendersn.rename(columns={"gender":"percent"}, inplace=True)
gendersn.head()

proportion
occupation    gender            
administrator M        54.430380
              F        45.569620
artist        M        53.571429
              F        46.428571
doctor        M       100.000000

In [16]:
new_gender = pd.DataFrame(gendersn.reset_index(level=["occupation", "gender"]))
new_gender.head()


occupation gender  proportion
0  administrator      M   54.430380
1  administrator      F   45.569620
2         artist      M   53.571429
3         artist      F   46.428571
4         doctor      M  100.000000

step6 for each occupation, calculate the minimum and maximum ages


In [26]:
min_max_ages_per_occupation = users.groupby('occupation').age.agg(['min', 'max'])
print("\nMinimum and maximum ages per occupation:")
print(min_max_ages_per_occupation)


Minimum and maximum ages per occupation:
               min  max
occupation             
administrator   21   70
artist          19   48
doctor          28   64
educator        23   63
engineer        22   70
entertainment   15   50
executive       22   69
healthcare      22   62
homemaker       20   50
lawyer          21   53
librarian       23   69
marketing       24   55
none            11   55
other           13   64
programmer      20   63
retired         51   73
salesman        18   66
scientist       23   55
student          7   42
technician      21   55
writer          18   60


step 7. for each combination of occupation and sex, calculate the mean age


In [27]:
users.groupby(['occupation', 'gender']).age.mean()

occupation     gender
administrator  F         40.638889
               M         37.162791
artist         F         30.307692
               M         32.333333
doctor         M         43.571429
educator       F         39.115385
               M         43.101449
engineer       F         29.500000
               M         36.600000
entertainment  F         31.000000
               M         29.000000
executive      F         44.000000
               M         38.172414
healthcare     F         39.818182
               M         45.400000
homemaker      F         34.166667
               M         23.000000
lawyer         F         39.500000
               M         36.200000
librarian      F         40.000000
               M         40.000000
marketing      F         37.200000
               M         37.875000
none           F         36.500000
               M         18.600000
other          F         35.472222
               M         34.028986
programmer     F         32.16666

step 8. for each occupation present the percentage of women and men

In [28]:
# create a DataFrame and apply count to gender
gender_ocup = users.groupby(['occupation', 'gender']).agg({'gender': 'count'})
# create a DataFrame and apply count for each occupation
occup_count = users.groupby(['occupation']).count()
# divide the gender_ocup per the occup_count and multiply per 100
occup_gender = gender_ocup.div(occup_count, level = "occupation")
occup_gender.loc[:, 'gender']

occupation     gender
administrator  F         0.455696
               M         0.544304
artist         F         0.464286
               M         0.535714
doctor         M         1.000000
educator       F         0.273684
               M         0.726316
engineer       F         0.029851
               M         0.970149
entertainment  F         0.111111
               M         0.888889
executive      F         0.093750
               M         0.906250
healthcare     F         0.687500
               M         0.312500
homemaker      F         0.857143
               M         0.142857
lawyer         F         0.166667
               M         0.833333
librarian      F         0.568627
               M         0.431373
marketing      F         0.384615
               M         0.615385
none           F         0.444444
               M         0.555556
other          F         0.342857
               M         0.657143
programmer     F         0.090909
               M         0

## Question 2

In [29]:
import pandas as pd

Step 2&3

In [30]:
euro12 = pd.read_csv('https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/02_Filtering_%26_Sorting/Euro12/Euro_2012_stats_TEAM.csv')
euro12.head()

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
0         Croatia      4               13                12             51.9%   
1  Czech Republic      4               13                18             41.9%   
2         Denmark      4               10                10             50.0%   
3         England      5               11                18             50.0%   
4          France      3               22                24             37.9%   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
0            16.0%                          32             0              0   
1            12.9%                          39             0              0   
2            20.0%                          27             1              0   
3            17.2%                          40             0              0   
4             6.5%                          65             1              0   

   Penalties not scored  ...  Saves made  Saves-to-shots ratio  Fouls Won  \
0                     0  ...          13                 81.3%         41   
1                     0  ...           9                 60.1%         53   
2                     0  ...          10                 66.7%         25   
3                     0  ...          22                 88.1%         43   
4                     0  ...           6                 54.6%         36   

  Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  Subs off  \
0             62         2             9          0        9         9   
1             73         8             7          0       11        11   
2             38         8             4          0        7         7   
3             45         6             5          0       11        11   
4             51         5             6          0       11        11   

   Players Used  
0            16  
1            19  
2            15  
3            16  
4            19  

[5 rows x 35 columns]

step 4

In [32]:
euro12[['Goals']]

Goals
0       4
1       4
2       4
3       5
4       3
5      10
6       5
7       6
8       2
9       2
10      6
11      1
12      5
13     12
14      5
15      2

Step 5.

In [33]:
euro12['Team'].count()

16

step 6.

In [34]:
euro12.shape[1]


35

step 7.

In [35]:
discipline = pd.DataFrame(euro12, columns=['Team','Yellow Cards', 'Red Cards'])
discipline

Team  Yellow Cards  Red Cards
0               Croatia             9          0
1        Czech Republic             7          0
2               Denmark             4          0
3               England             5          0
4                France             6          0
5               Germany             4          0
6                Greece             9          1
7                 Italy            16          0
8           Netherlands             5          0
9                Poland             7          1
10             Portugal            12          0
11  Republic of Ireland             6          1
12               Russia             6          0
13                Spain            11          0
14               Sweden             7          0
15              Ukraine             5          0

step 8

In [37]:
discipline.sort_values(by = ['Red Cards','Yellow Cards'], inplace = True)
discipline

Team  Yellow Cards  Red Cards
2               Denmark             4          0
5               Germany             4          0
3               England             5          0
8           Netherlands             5          0
15              Ukraine             5          0
4                France             6          0
12               Russia             6          0
1        Czech Republic             7          0
14               Sweden             7          0
0               Croatia             9          0
13                Spain            11          0
10             Portugal            12          0
7                 Italy            16          0
11  Republic of Ireland             6          1
9                Poland             7          1
6                Greece             9          1

Step 9.

In [38]:
euro12.groupby('Team')['Yellow Cards'].mean()

Team
Croatia                 9.0
Czech Republic          7.0
Denmark                 4.0
England                 5.0
France                  6.0
Germany                 4.0
Greece                  9.0
Italy                  16.0
Netherlands             5.0
Poland                  7.0
Portugal               12.0
Republic of Ireland     6.0
Russia                  6.0
Spain                  11.0
Sweden                  7.0
Ukraine                 5.0
Name: Yellow Cards, dtype: float64

step 10.

In [39]:
euro12[euro12['Goals'] > 6]

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
5   Germany     10               32                32             47.8%   
13    Spain     12               42                33             55.9%   

   % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
5             15.6%                          80             2              1   
13            16.0%                         100             0              1   

    Penalties not scored  ...  Saves made  Saves-to-shots ratio  Fouls Won  \
5                      0  ...          10                 62.6%         63   
13                     0  ...          15                 93.8%        102   

   Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  Subs off  \
5              49        12             4          0       15        15   
13             83        19            11          0       17        17   

    Players Used  
5             17  
13            18  

[2 rows x 35 columns]

Step:11

In [40]:
euro12[euro12.Team.str.startswith('G')]

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
5  Germany     10               32                32             47.8%   
6   Greece      5                8                18             30.7%   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
5            15.6%                          80             2              1   
6            19.2%                          32             1              1   

   Penalties not scored  ...  Saves made  Saves-to-shots ratio  Fouls Won  \
5                     0  ...          10                 62.6%         63   
6                     1  ...          13                 65.1%         67   

  Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  Subs off  \
5             49        12             4          0       15        15   
6             48        12             9          1       12        12   

   Players Used  
5            17  
6            20  

[2 rows x 35 columns]

step 12.

In [41]:
euro12.iloc[ : , : 7]

Team  Goals  Shots on target  Shots off target  \
0               Croatia      4               13                12   
1        Czech Republic      4               13                18   
2               Denmark      4               10                10   
3               England      5               11                18   
4                France      3               22                24   
5               Germany     10               32                32   
6                Greece      5                8                18   
7                 Italy      6               34                45   
8           Netherlands      2               12                36   
9                Poland      2               15                23   
10             Portugal      6               22                42   
11  Republic of Ireland      1                7                12   
12               Russia      5                9                31   
13                Spain     12               42                33   
14               Sweden      5               17                19   
15              Ukraine      2                7                26   

   Shooting Accuracy % Goals-to-shots  Total shots (inc. Blocked)  
0              51.9%            16.0%                          32  
1              41.9%            12.9%                          39  
2              50.0%            20.0%                          27  
3              50.0%            17.2%                          40  
4              37.9%             6.5%                          65  
5              47.8%            15.6%                          80  
6              30.7%            19.2%                          32  
7              43.0%             7.5%                         110  
8              25.0%             4.1%                          60  
9              39.4%             5.2%                          48  
10             34.3%             9.3%                          82  
11             36.8%             5.2%                          28  
12             22.5%            12.5%                          59  
13             55.9%            16.0%                         100  
14             47.2%            13.8%                          39  
15             21.2%             6.0%                          38

step 13.

In [43]:
euro12.iloc[ : , : -3]

Team  Goals  Shots on target  Shots off target  \
0               Croatia      4               13                12   
1        Czech Republic      4               13                18   
2               Denmark      4               10                10   
3               England      5               11                18   
4                France      3               22                24   
5               Germany     10               32                32   
6                Greece      5                8                18   
7                 Italy      6               34                45   
8           Netherlands      2               12                36   
9                Poland      2               15                23   
10             Portugal      6               22                42   
11  Republic of Ireland      1                7                12   
12               Russia      5                9                31   
13                Spain     12               42                33   
14               Sweden      5               17                19   
15              Ukraine      2                7                26   

   Shooting Accuracy % Goals-to-shots  Total shots (inc. Blocked)  \
0              51.9%            16.0%                          32   
1              41.9%            12.9%                          39   
2              50.0%            20.0%                          27   
3              50.0%            17.2%                          40   
4              37.9%             6.5%                          65   
5              47.8%            15.6%                          80   
6              30.7%            19.2%                          32   
7              43.0%             7.5%                         110   
8              25.0%             4.1%                          60   
9              39.4%             5.2%                          48   
10             34.3%             9.3%                          82   
11             36.8%             5.2%                          28   
12             22.5%            12.5%                          59   
13             55.9%            16.0%                         100   
14             47.2%            13.8%                          39   
15             21.2%             6.0%                          38   

    Hit Woodwork  Penalty goals  Penalties not scored  ...  Clean Sheets  \
0              0              0                     0  ...             0   
1              0              0                     0  ...             1   
2              1              0                     0  ...             1   
3              0              0                     0  ...             2   
4              1              0                     0  ...             1   
5              2              1                     0  ...             1   
6              1              1                     1  ...             1   
7              2              0                     0  ...             2   
8              2              0                     0  ...             0   
9              0              0                     0  ...             0   
10             6              0                     0  ...             2   
11             0              0                     0  ...             0   
12             2              0                     0  ...             0   
13             0              1                     0  ...             5   
14             3              0                     0  ...             1   
15             0              0                     0  ...             0   

    Blocks  Goals conceded Saves made  Saves-to-shots ratio  Fouls Won  \
0       10               3         13                 81.3%         41   
1       10               6          9                 60.1%         53   
2       10               5         10                 66.7%         25   
3       29               3         22                 88.1%         43   
4       

In [44]:
#step14 - present only the shooting accuracy from England, Italy and Russia
choosen_teams = ['England', 'Italy', 'Russia']
shooting_accuracy_choosen_teams = euro12[euro12['Team'].isin(choosen_teams)][['Team', 'Shooting Accuracy']]
print(shooting_accuracy_choosen_teams)

       Team Shooting Accuracy
3   England             50.0%
7     Italy             43.0%
12   Russia             22.5%


Question 3

Housing Step 1. Import the necessary libraries Step 2. Create 3 differents Series, each of length 100, as follows: • The first a random number from 1 to 4 • The second a random number from 1 to 3 • The third a random number from 10,000 to 30,000 Step 3. Create a DataFrame by joinning the Series by column Step 4. Change the name of the columns to bedrs, bathrs, price_sqr_meter Step 5. Create a one column DataFrame with the values of the 3 Series and assign it to 'bigcolumn' Step 6. Ops it seems it is going only until index 99. Is it true? Step 7. Reindex the DataFrame so it goes from 0 to 299

In [45]:
#Step1 - import libraries
import pandas as pd
import random


In [46]:
#step2 - create 3 different series, each of length 100
#create dataframes as stated
first = [[random.randint(1, 4)] for i in range(100)]
df_1 = pd.DataFrame(first)

second = [[random.randint(1, 3)] for i in range(100)]
df_2 = pd.DataFrame(second)

third = [[random.randint(10000, 30000)] for i in range(100)]
df_3 = pd.DataFrame(third)

In [47]:
#step3 - create DataFrame by joining the series by column
df = pd.concat([df_1, df_2, df_3], axis=1)
df

0  0      0
0   1  2  29329
1   3  1  17288
2   3  3  19488
3   4  2  12295
4   3  2  22102
.. .. ..    ...
95  3  2  21322
96  2  3  25163
97  2  3  17095
98  2  1  15102
99  2  1  15006

[100 rows x 3 columns]

In [48]:
#step4 - change the name of columns
df.columns = ['bedrs', 'bathrs', 'price_sqr_meter']
df

bedrs  bathrs  price_sqr_meter
0       1       2            29329
1       3       1            17288
2       3       3            19488
3       4       2            12295
4       3       2            22102
..    ...     ...              ...
95      3       2            21322
96      2       3            25163
97      2       3            17095
98      2       1            15102
99      2       1            15006

[100 rows x 3 columns]

In [49]:
#step5 - create a one colum DataFrame with the values of the 3series and asign it to 'bigcolumn'
dataFrame_new = df.bedrs.astype(str).str.cat(df.bathrs.astype(str)).str.cat(df.price_sqr_meter.astype(str))
dataFrame_new.columns = ['bigcolumn']
dataFrame_new

0     1229329
1     3117288
2     3319488
3     4212295
4     3222102
       ...   
95    3221322
96    2325163
97    2317095
98    2115102
99    2115006
Name: bedrs, Length: 100, dtype: object

In [50]:
#step6 - checking if it goes only until index 99
print("Is it true that bigcolumn goes only until index 99 :", dataFrame_new.index.max() == 99)

Is it true that bigcolumn goes only until index 99 : True


In [51]:
#step7 - reindex the DataFrame so it goes from 0 to 299
dataFrame_new = dataFrame_new.reset_index(drop=True)
dataFrame_new.reset_index()
dataFrame_new.reindex(index=range(0,300))

0      1229329
1      3117288
2      3319488
3      4212295
4      3222102
        ...   
295        NaN
296        NaN
297        NaN
298        NaN
299        NaN
Name: bedrs, Length: 300, dtype: object

Question 4

Wind Statistics
The data have been modified to contain some missing values, identified by NaN. Using pandas should make this exercise easier, in particular for the bonus question. You should be able to perform all of these operations without using a for loop or other looping construct. The data in 'wind.data' has the following format: Yr Mo Dy RPT VAL ROS KIL SHA BIR DUB CLA MUL CLO BEL MAL 61 1 1 15.04 14.96 13.17 9.29 NaN 9.87 13.67 10.25 10.83 12.58 18.50 15.04 61 1 2 14.71 NaN 10.83 6.50 12.62 7.67 11.50 10.04 9.79 9.67 17.54 13.83 61 1 3 18.50 16.88 12.33 10.13 11.17 6.17 11.25 NaN 8.50 7.67 12.75 12.71

The first three columns are year, month, and day. The remaining 12 columns are average windspeeds in knots at 12 locations in Ireland on that day.
Step 1. Import the necessary libraries Step 2. Import the dataset from the attached file wind.txt Step 3. Assign it to a variable called data and replace the first 3 columns by a proper datetime index. Step 4. Year 2061? Do we really have data from this year? Create a function to fix it and apply it. Step 5. Set the right dates as the index. Pay attention at the data type, it should be datetime64[ns]. Step 6. Compute how many values are missing for each location over the entire record.They should be ignored in all calculations below. Step 7. Compute how many non-missing values there are in total. Step 8. Calculate the mean windspeeds of the windspeeds over all the locations and all the times. A single number for the entire dataset. Step 9. Create a DataFrame called loc_stats and calculate the min, max and mean windspeeds and standard deviations of the windspeeds at each location over all the days A different set of numbers for each location. Step 10. Create a DataFrame called day_stats and calculate the min, max and mean windspeed and standard deviations of the windspeeds across all the locations at each day. A different set of numbers for each day. Step 11. Find the average windspeed in January for each location. Treat January 1961 and January 1962 both as January. Step 12. Downsample the record to a yearly frequency for each location. Step 13. Downsample the record to a monthly frequency for each location. Step 14. Downsample the record to a weekly frequency for each location. Step 15. Calculate the min, max and mean windspeeds and standard deviations of the windspeeds across all locations for each week (assume that the first week starts on January 2 1961) for the first 52 weeks.

In [59]:
#step 1 - importing the necessary libraries
import pandas as pd
import numpy as np


In [60]:
#step 2 - import the dataset
data = pd.read_csv('https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/06_Stats/Wind_Stats/wind.data', sep = '\s+')
data.tail()

Yr  Mo  Dy    RPT    VAL    ROS   KIL    SHA    BIR    DUB    CLA  \
6569  78  12  27  17.58  16.96  17.62  8.08  13.21  11.67  14.46  15.59   
6570  78  12  28  13.21   5.46  13.46  5.00   8.12   9.42  14.33  16.25   
6571  78  12  29  14.00  10.29  14.42  8.71   9.71  10.54  19.17  12.46   
6572  78  12  30  18.50  14.04  21.29  9.13  12.75   9.71  18.08  12.87   
6573  78  12  31  20.33  17.41  27.29  9.59  12.08  10.13  19.25  11.63   

        MUL    CLO    BEL    MAL  
6569  14.04  14.00  17.21  40.08  
6570  15.25  18.05  21.79  41.46  
6571  14.50  16.42  18.88  29.58  
6572  12.46  12.12  14.67  28.79  
6573  11.58  11.38  12.08  22.08

In [61]:
data["Date"] = pd.to_datetime(data[["Yr","Mo","Dy"]].astype(str).agg('-'.join, axis=1))
data = data.drop(columns=["Yr","Mo","Dy"])
#re-arrange cols
data = data[['Date','RPT','VAL','ROS','KIL','SHA','BIR','DUB','CLA','MUL','CLO','BEL','MAL']]
data.head()

C:\Users\ravir\AppData\Local\Temp\ipykernel_9372\4099077853.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data["Date"] = pd.to_datetime(data[["Yr","Mo","Dy"]].astype(str).agg('-'.join, axis=1))


Date    RPT    VAL    ROS    KIL    SHA   BIR    DUB    CLA    MUL  \
0 2061-01-01  15.04  14.96  13.17   9.29    NaN  9.87  13.67  10.25  10.83   
1 2061-01-02  14.71    NaN  10.83   6.50  12.62  7.67  11.50  10.04   9.79   
2 2061-01-03  18.50  16.88  12.33  10.13  11.17  6.17  11.25    NaN   8.50   
3 2061-01-04  10.58   6.63  11.75   4.58   4.54  2.88   8.63   1.79   5.83   
4 2061-01-05  13.33  13.25  11.42   6.17  10.71  8.21  11.92   6.54  10.92   

     CLO    BEL    MAL  
0  12.58  18.50  15.04  
1   9.67  17.54  13.83  
2   7.67  12.75  12.71  
3   5.88   5.46  10.88  
4  10.34  12.92  11.83

In [62]:
#step 4 - Year 2061? Do we really have data from this year? Create a function to fix it and apply it
data["Date"] = np.where(pd.DatetimeIndex(data["Date"]).year < 2000,
                        data.Date,data.Date - pd.offsets.DateOffset(years=100))
data.head()

Date    RPT    VAL    ROS    KIL    SHA   BIR    DUB    CLA    MUL  \
0 1961-01-01  15.04  14.96  13.17   9.29    NaN  9.87  13.67  10.25  10.83   
1 1961-01-02  14.71    NaN  10.83   6.50  12.62  7.67  11.50  10.04   9.79   
2 1961-01-03  18.50  16.88  12.33  10.13  11.17  6.17  11.25    NaN   8.50   
3 1961-01-04  10.58   6.63  11.75   4.58   4.54  2.88   8.63   1.79   5.83   
4 1961-01-05  13.33  13.25  11.42   6.17  10.71  8.21  11.92   6.54  10.92   

     CLO    BEL    MAL  
0  12.58  18.50  15.04  
1   9.67  17.54  13.83  
2   7.67  12.75  12.71  
3   5.88   5.46  10.88  
4  10.34  12.92  11.83

In [63]:
#step 5 - Set the right dates as the index. Pay attention at the data type, it should be datetime64[ns].
NData = data.set_index("Date")
NData.index.astype("datetime64[ns]")

DatetimeIndex(['1961-01-01', '1961-01-02', '1961-01-03', '1961-01-04',
               '1961-01-05', '1961-01-06', '1961-01-07', '1961-01-08',
               '1961-01-09', '1961-01-10',
               ...
               '1978-12-22', '1978-12-23', '1978-12-24', '1978-12-25',
               '1978-12-26', '1978-12-27', '1978-12-28', '1978-12-29',
               '1978-12-30', '1978-12-31'],
              dtype='datetime64[ns]', name='Date', length=6574, freq=None)

In [64]:
#step 6 - compute how many values are missing for each location over the entire record. They should be ignored in all calculations below
NData.isnull().values.ravel().sum()

31

In [65]:
#step 7 - compute how many non-missing values there are in total
nData = NData.count()
print("Total Non-missing values is :", nData.sum())

Total Non-missing values is : 78857


In [66]:
#step 8 - Calculate the mean windspeeds of the windspeeds over all the locations and all the times.
a = round(NData.mean(),2)
a

RPT    12.36
VAL    10.64
ROS    11.66
KIL     6.31
SHA    10.46
BIR     7.09
DUB     9.80
CLA     8.50
MUL     8.49
CLO     8.71
BEL    13.12
MAL    15.60
dtype: float64

In [67]:
#A single number for the entire data set
round(a.mean(),2)

10.23

In [68]:
#step 9 - Create a DataFrame called loc_stats and calculate the min, max and mean windspeeds and standard deviations of the windspeeds at each location over all the days A different set of numbers for each location. 
def statsKPI(a):
    a = pd.Series(a)
    Mean = a.mean()
    Std = a.std()
    Min = a.min()
    Max = a.max()
    res = [Min,Max,Mean,Std]
    indx = ["MEAN","Std","Min","Max"]
    result = pd.Series(res,index=indx)
    return result

In [69]:
#different set of numbers for each location.
loc_stats = NData.apply(statsKPI)
loc_stats

RPT        VAL        ROS        KIL        SHA        BIR  \
MEAN   0.670000   0.210000   1.500000   0.000000   0.130000   0.000000   
Std   35.800000  33.370000  33.840000  28.460000  37.540000  26.160000   
Min   12.362987  10.644314  11.660526   6.306468  10.455834   7.092254   
Max    5.618413   5.267356   5.008450   3.605811   4.936125   3.968683   

            DUB        CLA        MUL        CLO        BEL        MAL  
MEAN   0.000000   0.000000   0.000000   0.040000   0.130000   0.670000  
Std   30.370000  31.080000  25.880000  28.210000  42.380000  42.540000  
Min    9.797343   8.495053   8.493590   8.707332  13.121007  15.599079  
Max    4.977555   4.499449   4.166872   4.503954   5.835037   6.699794

In [70]:
#Step 10 - Create a DataFrame called day_stats and calculate the min, max and mean windspeed and standard deviations of the windspeeds across all the locations at each day. A different set of numbers for each day.
day_stats = NData.apply(statsKPI,axis=1)
day_stats.tail()

MEAN    Std        Min       Max
Date                                        
1978-12-27  8.08  40.08  16.708333  7.868076
1978-12-28  5.00  41.46  15.150000  9.687857
1978-12-29  8.71  29.58  14.890000  5.756836
1978-12-30  9.13  28.79  15.367500  5.540437
1978-12-31  9.59  27.29  15.402500  5.702483

In [71]:
#step 11 - Find the average windspeed in January for each location. Treat January 1961 and January 1962 both as January.
january_AVG_windspeed = NData[NData.index.month == 1]
round(january_AVG_windspeed.mean(),2)

RPT    14.85
VAL    12.91
ROS    13.30
KIL     7.20
SHA    11.67
BIR     8.05
DUB    11.82
CLA     9.51
MUL     9.54
CLO    10.05
BEL    14.55
MAL    18.03
dtype: float64

In [72]:
#step 12 - Downsample the record to a yearly frequency for each location
print( "Yearly Frequency:\n", NData.resample('A').mean())

Yearly Frequency:
                   RPT        VAL        ROS       KIL        SHA       BIR  \
Date                                                                         
1961-12-31  12.299583  10.351796  11.362369  6.958227  10.881763  7.729726   
1962-12-31  12.246923  10.110438  11.732712  6.960440  10.657918  7.393068   
1963-12-31  12.813452  10.836986  12.541151  7.330055  11.724110  8.434712   
1964-12-31  12.363661  10.920164  12.104372  6.787787  11.454481  7.570874   
1965-12-31  12.451370  11.075534  11.848767  6.858466  11.024795  7.478110   
1966-12-31  13.461973  11.557205  12.020630  7.345726  11.805041  7.793671   
1967-12-31  12.737151  10.990986  11.739397  7.143425  11.630740  7.368164   
1968-12-31  11.835628  10.468197  11.409754  6.477678  10.760765  6.067322   
1969-12-31  11.166356   9.723699  10.902000  5.767973   9.873918  6.189973   
1970-12-31  12.600329  10.726932  11.730247  6.217178  10.567370  7.609452   
1971-12-31  11.273123   9.095178  11.088329  

In [73]:
#step 13 - Downsample the record to a monthly frequency for each location
print("Monthly Frequency:\n", NData.resample('M').mean())

Monthly Frequency:
                   RPT        VAL        ROS       KIL        SHA        BIR  \
Date                                                                          
1961-01-31  14.841333  11.988333  13.431613  7.736774  11.072759   8.588065   
1961-02-28  16.269286  14.975357  14.441481  9.230741  13.852143  10.937500   
1961-03-31  10.890000  11.296452  10.752903  7.284000  10.509355   8.866774   
1961-04-30  10.722667   9.427667   9.998000  5.830667   8.435000   6.495000   
1961-05-31   9.860968   8.850000  10.818065  5.905333   9.490323   6.574839   
...               ...        ...        ...       ...        ...        ...   
1978-08-31   9.645161   8.259355   9.032258  4.502903   7.368065   5.935161   
1978-09-30  10.913667  10.895000  10.635000  5.725000  10.372000   9.278333   
1978-10-31   9.897742   8.670968   9.295806  4.721290   8.525161   6.774194   
1978-11-30  16.151667  14.802667  13.508000  7.317333  11.475000   8.743000   
1978-12-31  16.175484  13.748065

In [74]:
#step 14 - Downsample the record to a weekly frequency for each location
print("Weekly Frequency:\n", NData.resample('W').mean())

Weekly Frequency:
                   RPT        VAL        ROS        KIL        SHA        BIR  \
Date                                                                           
1961-01-01  15.040000  14.960000  13.170000   9.290000        NaN   9.870000   
1961-01-08  13.541429  11.486667  10.487143   6.417143   9.474286   6.435714   
1961-01-15  12.468571   8.967143  11.958571   4.630000   7.351429   5.072857   
1961-01-22  13.204286   9.862857  12.982857   6.328571   8.966667   7.417143   
1961-01-29  19.880000  16.141429  18.225714  12.720000  17.432857  14.828571   
...               ...        ...        ...        ...        ...        ...   
1978-12-03  14.934286  11.232857  13.941429   5.565714  10.215714   8.618571   
1978-12-10  20.740000  19.190000  17.034286   9.777143  15.287143  12.774286   
1978-12-17  16.758571  14.692857  14.987143   6.917143  11.397143   7.272857   
1978-12-24  11.155714   8.008571  13.172857   4.004286   7.825714   6.290000   
1978-12-31  14.951429

In [75]:
#step 15 - Calculate the min, max and mean windspeeds and standard deviations of the windspeeds across all locations for each week (assume that the first week starts on January 2 1961) for the first 52 weeks.
statistic1 = NData.resample('W').mean().apply(lambda x: [x.count(),x.min(),x.max(),x.mean(),x.std()])
print(statistic1)

          RPT         VAL         ROS         KIL         SHA         BIR  \
0  940.000000  940.000000  940.000000  940.000000  939.000000  940.000000   
1    3.874286    2.452857    4.901429    1.447143    2.775714    1.382857   
2   24.010000   23.715714   21.384286   14.517143   22.897143   16.010000   
3   12.365304   10.647725   11.661255    6.309334   10.455607    7.094787   
4    3.466600    3.372501    2.840361    2.186432    3.146715    2.556633   

          DUB         CLA         MUL         CLO         BEL         MAL  
0  940.000000  940.000000  940.000000  940.000000  940.000000  940.000000  
1    2.471429    2.284286    2.262857    1.571429    3.661429    5.124286  
2   20.474286   17.861429   17.460000   17.932857   26.617143   30.374286  
3    9.801271    8.496123    8.496452    8.710639   13.125912   15.597171  
4    3.308180    2.853029    2.638349    2.924546    3.757189    4.468526  


In [76]:
first_yr = NData[NData.index.year == 1961]
statistic1 = NData.resample('W').mean().apply(lambda x: x.describe())
print(statistic1)

              RPT         VAL         ROS         KIL         SHA         BIR  \
count  940.000000  940.000000  940.000000  940.000000  939.000000  940.000000   
mean    12.365304   10.647725   11.661255    6.309334   10.455607    7.094787   
std      3.466600    3.372501    2.840361    2.186432    3.146715    2.556633   
min      3.874286    2.452857    4.901429    1.447143    2.775714    1.382857   
25%      9.884286    8.277857    9.613929    4.742857    8.248571    5.211071   
50%     12.291429   10.486429   11.490000    6.165714   10.254286    6.933571   
75%     14.547143   12.757857   13.457143    7.584286   12.627143    8.842500   
max     24.010000   23.715714   21.384286   14.517143   22.897143   16.010000   

              DUB         CLA         MUL         CLO         BEL         MAL  
count  940.000000  940.000000  940.000000  940.000000  940.000000  940.000000  
mean     9.801271    8.496123    8.496452    8.710639   13.125912   15.597171  
std      3.308180    2.853029 

Question 5

Step 1. Import the necessary libraries Step 2. Import the dataset from this address. Step 3. Assign it to a variable called chipo. Step 4. See the first 10 entries Step 5. What is the number of observations in the dataset? Step 6. What is the number of columns in the dataset? Step 7. Print the name of all the columns. Step 8. How is the dataset indexed? Step 9. Which was the most-ordered item? Step 10. For the most-ordered item, how many items were ordered? Step 11. What was the most ordered item in the choice_description column? Step 12. How many items were orderd in total? Step 13. • Turn the item price into a float • Check the item price type • Create a lambda function and change the type of item price • Check the item price type Step 14. How much was the revenue for the period in the dataset? Step 15. How many orders were made in the period? Step 16. What is the average revenue amount per order? Step 17. How many different items are sold?

In [78]:
#step 1 - import the necessary library
import pandas as pd

In [79]:
#step 2 - import the dataset
#step 3 - assign it to a variable called chipo
data_url = "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv"
chipo = pd.read_csv(data_url, sep='\t')

In [80]:
#step 4 - first 10 entries
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

In [81]:
#step 5 - What is the number of observation in the dataset?
print ("\nNumber of observation in the dataset is", chipo.shape[0])


Number of observation in the dataset is 4622


In [82]:
#step 6 - number of columns in the dataset
print("\nNumber of columns in the dataset is", chipo.shape[1])


Number of columns in the dataset is 5


In [83]:
#step 7 - print the name of all the columns
chipo.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

In [84]:
#step 8 - how is the dataset indexed?
chipo.index

RangeIndex(start=0, stop=4622, step=1)

In [85]:
#step 9 - which was the most ordered item?
item = chipo['item_name'].mode().iloc[0]
print(item)

Chicken Bowl


In [86]:
#step 10 - for the most ordered, how many items were ordered?
item = chipo[chipo['item_name'] == item]['quantity'].sum()
print(item)

761


In [87]:
#step 11 - What was the most ordered item in the choice_description column?
most_ordered_choice = chipo['choice_description'].mode().iloc[0]
print("\nMost ordered item choice description is", most_ordered_choice)


Most ordered item choice description is [Diet Coke]


In [88]:
#step 12 - How many items were ordered in total
chipo.quantity.sum()

4972

In [93]:
#step 13 - Turn the item price into a float • Check the item price type • Create a lambda function and change the type of item price • Check the item price type 
#item price type
chipo.item_price.dtype

dtype('float64')

In [97]:

chipo.item_price.dtype

dtype('float64')

In [98]:
#step 14 - How much was the revenue for the period in the dataset?/
#revenue for a dataset
revenue = (chipo['quantity']* chipo['item_price']).sum()
print("\nRevenue for the period in dataset:", revenue)


Revenue for the period in dataset: 39237.02


In [99]:
#step 15 -total orders made
number_orders = chipo.order_id.value_counts().count()
print(number_orders)

1834


In [100]:
#step 16 - average revenue per order
avg_rev_per_order = revenue / number_orders
print(avg_rev_per_order)

21.39423118865867


In [101]:
#step 17 - How many different items are sold?
chipo.item_name.value_counts().count()

50